[Clang] Add __builtin_counted_by_ref builtin

bwendling opened the pull request:
The __builtin_counted_by_ref builtin is used on a flexible array
pointer and returns a pointer to the "counted_by" attribute's COUNT
argument, which is a field in the same non-anonymous struct as the
flexible array member. This is useful for automatically setting the
count field without needing the programmer's intervention. Otherwise
it's possible to get this anti-pattern:
    
      ptr = alloc(<ty>, ..., COUNT);
      ptr->FAM[9] = 42; /* <<< Sanitizer will complain */
      ptr->count = COUNT;
    
To prevent this anti-pattern, the user can create an allocator that
automatically performs the assignment:
    
      #define alloc(TY, FAM, COUNT) ({ \
          TY __p = alloc(get_size(TY, COUNT));             \
          if (__builtin_counted_by_ref(__p->FAM))          \
              *__builtin_counted_by_ref(__p->FAM) = COUNT; \
          __p;                                             \
      })

The builtin's behavior is heavily dependent upon the "counted_by"
attribute existing. It's main utility is during allocation to avoid
the above anti-pattern. If the flexible array member doesn't have that
attribute, the builtin becomes a no-op. Therefore, if the flexible
array member has a "count" field not referenced by "counted_by", it
must be set explicitly after the allocation as this builtin will
return a "nullptr" and the assignment will most likely be elided.

bwendling committed:
[Clang] Add __builtin_counted_by_ref builtin

The __builtin_counted_by_ref builtin is used on a flexible array
pointer and returns a pointer to the "counted_by" attribute's COUNT
argument, which is a field in the same non-anonymous struct as the
flexible array member. This is useful for automatically setting the
count field without needing the programmer's intervention. Otherwise
it's possible to get this anti-pattern:

  ptr = alloc(<ty>, ..., COUNT);
  ptr->FAM[9] = 42; /* <<< Sanitizer will complain */
  ptr->count = COUNT;

To prevent this anti-pattern, the user can create an allocator that
automatically performs the assignment:

  #define alloc(TY, FAM, COUNT) ({ \
      TY __p = alloc(get_size(TY, COUNT));             \
      if (__builtin_counted_by_ref(__p->FAM))          \
          *__builtin_counted_by_ref(__p->FAM) = COUNT; \
      __p;                                             \
  })
The builtin's behavior is heavily dependent upon the "counted_by"
attribute existing. It's main utility is during allocation to avoid
the above anti-pattern. If the flexible array member doesn't have that
attribute, the builtin becomes a no-op. Therefore, if the flexible
array member has a "count" field not referenced by "counted_by", it
must be set explicitly after the allocation as this builtin will
return a "nullptr" and the assignment will most likely be elided.

bwendling committed:
If the 'counted_by' attribute doesn't exist, then return nullptr.

bwendling committed:
Remove debugging code.

bwendling committed:
Make returning a nullptr the default.

bwendling committed:
clang-format

bwendling committed:
Use a visitor to retrieve the MemberExpr. It's intentionally harsh and ignores pretty much everythings that may 'hide' the underlying MemberExpr.

bwendling committed:
Add testcases.

bwendling committed:
Combine the code to grab the MemberExpr, if it exists.

bwendling committed:
Move FindCountedByField into a central place.

bwendling committed:
Use CustomTypeChecking for the builtin. It allows us explicitly to set the return type based on the 'count' field's type.

bwendling committed:
[Clang][NFC] Move FindCountedByField into FieldDecl (#104235)

FindCountedByField can be used in more places than CodeGen. Move it into
FieldDecl to avoid layering issues.

bwendling committed:
Fix a build failure.

bwendling committed:
Remove empty line.

bwendling committed:
Remove merge errors.

bwendling committed:
Use 'size_t *' for the default return type and improve the testcase.

bwendling committed:
Don't use a Visitor to retrieve the MemberExpr. Instead accept only non-pathological expressions.

bwendling committed:
Improve error messages. Make the side-effects error a warning.

bwendling committed:
Add _Generic checks.

bwendling committed:
Remove unrelated change

bwendling committed:
Use MemberExpr::Create instead of MemberExpr::CreateImplicit.

bwendling committed:
Regenerate

bwendling committed:
Add -W flag for warning.

bwendling committed:
Handle any anonymous structs surrounding the flexible array member and / or count field.

bwendling committed:
Add anonymous struct testcase.

bwendling committed:
Report a fatal error if we can't find the 'count' field.

isanbard committed:
Use 'cast' instead of 'dyn_cast' and default to 'void *' for the return type.

bwendling committed:
A 'void *' can result in an error message about assigning to a 'void'. Go back to 'size_t *' as the default return type.

bwendling committed:
Add some documentation.

bwendling committed:
Rename to __builtin_counted_by_ref.

bwendling committed:
Rename method to reflect new builtin name.

bwendling committed:
Emit an error if the 'count' reference is assigned to a variable, used as a function arg, or in a return.

bwendling committed:
Ignore paren casts.

bwendling committed:
Don't allow __builtin_counted_by_ref in complex expressions.

bwendling committed:
Add test case for pointers.

bwendling committed:
Add SourceRange to the diagnostic.

bwendling committed:
Check that __builtin_counted_by_ref isn't used on the LHS by a complex equation.

bwendling committed:
Update clang/include/clang/Basic/DiagnosticSemaKinds.td

make the grammar better.

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

bwendling committed:
Add a flag indicating that the 'MemberExpr' was once a call to '__builtin_counted_by_ref' so that we can issue proper diagnostics.

bwendling committed:
Remove invalid tests that take the address of the bounds counter.

bwendling committed:
Add testcase that discards side-effects.

bwendling committed:
Delay warnings for _Generic statements.

bwendling committed:
Marking some diagnostics as 'potentially evaluated' somehow prevented variables from being marked as 'used'. >_<

bwendling committed:
Regenerate testcase.

bwendling committed:
Fix testcase.

bwendling committed:
Fix some of the naming and remove dead code.

bwendling committed:
Correct include ordering.

bwendling committed:
Improve the argument checking so that it's checking specifically for 'counted_by' and not other counting attributes. Also make the documentation better.

bwendling committed:
Move setting that the counter is a bounds-safety counter to where __builtin_counted_by_ref is processed so that future sema checks are concerned only with that code path.

bwendling committed:
Retain the __builtin_counted_by_ref CallExpr for AST dumping reasons. We'll lower it during CodeGen.

bwendling committed:
Fix

bwendling commented:
This is a new PR, because GitHub messed up. Here's the original: https://github.com/llvm/llvm-project/pull/102549

bwendling committed:
Reformat

bwendling committed:
Add correct attribute checking and set the return type.

bwendling committed:
Generate LLVM IR for the __builtin_counted_by_ref.

bwendling committed:
Set InBounds on the GEP.

bwendling committed:
Regenerate testcase.

bwendling committed:
Remove warning that no longer shows up.

bwendling committed:
Regenerate testcase

bwendling committed:
Expand testing. This includes a hack to prevent multiple diagnostics for the same issue.

bwendling committed:
Add test rejecting using of 'counted_by' in C++.

bwendling commented:
@AaronBallman @rapidsna -- I think I covered all of the comments you made on the other PR (stupid git!). PTAL.

AaronBallman reviewed (commented):
Can you add a test with `-ast-print` that shows we print out the builtin as it's used in the source, and a test with `-ast-dump` that shows the AST retains information about the use of the builtin?

bwendling committed:
Add -ast-print testcase.

bwendling commented:
-ast-print testcase added.

rapidsna commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -1182,12 +1174,25 @@ llvm::Value *CodeGenFunction::EmitLoadOfCountedByField(
-  Res = Builder.CreateInBoundsGEP(
-      ConvertType(QualType(RD->getTypeForDecl(), 0)), Res,
-      RecIndicesTy(llvm::reverse(Indices)), "..counted_by.gep");
+  return Builder.CreateGEP(ConvertType(QualType(RD->getTypeForDecl(), 0)), Res,
```
I think you still need `InBounds`. Is there a reason why `inbounds` is removed?

bwendling replied:
Looks like I might have caught a previous bogus commit here when I had to create a new PR because git is ass. I'll readd it.

bwendling committed:
Re-add inbounds to the GEP.

rapidsna commented on clang/test/CodeGen/attr-counted-by-pr110385.c:
```diff
@@ -31,7 +31,7 @@ void init(void * __attribute__((pass_dynamic_object_size(0))));
 // CHECK-NEXT:    [[GROWABLE:%.*]] = getelementptr inbounds nuw i8, ptr [[FOO]], i64 8
 // CHECK-NEXT:    [[TMP0:%.*]] = load ptr, ptr [[GROWABLE]], align 8, !tbaa [[TBAA2:![0-9]+]]
 // CHECK-NEXT:    [[ARRAY:%.*]] = getelementptr inbounds nuw i8, ptr [[TMP0]], i64 12
-// CHECK-NEXT:    [[DOT_COUNTED_BY_GEP:%.*]] = getelementptr inbounds i8, ptr [[TMP0]], i64 8
```
This seems to be a regression caused by removing `InBounds` from [here](https://github.com/llvm/llvm-project/pull/114495/files#r1828441628). Same for `attr-counted-by.c`.

bwendling replied:
Please do a refresh.

rapidsna commented on clang/test/Sema/builtin-counted-by-ref.c:
```diff
@@ -0,0 +1,120 @@
+}
+
+void test4(struct fam_struct *ptr, int idx) {
+  __builtin_counted_by_ref(&ptr->array[idx++]);             // expected-warning {{'__builtin_counted_by_ref' argument has side-effects that will be discarded}}
```
discarded means the argument is not evaluated? Could you please add a codegen test for this warning case (since it's still supposed to build and generate some code).

bwendling replied:
One already exists in tests/CodeGen/builtin-counted-by-ref.c `test4`.

rapidsna commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -5573,6 +5577,62 @@ bool Sema::BuiltinSetjmp(CallExpr *TheCall) {
+        << Arg->getSourceRange();
+
+  // See if we have something like '&ptr->fam[0]`.
+  if (auto *UO = dyn_cast<UnaryOperator>(Arg);
```
`&` and []` are treated as an optional that will be ignored if present, it means `__builtin_counted_by_ref(&p->array)` accidentally works? Should we make it illformed if `AddrOf` is not followed by `ArraySubscriptExpr`? 

And what about `&ptr->fam[2]`?

bwendling replied:
I have a test for `&ptr->array[idx]`. I'll add one for `&ptr->array`.

bwendling committed:
Add test for &ptr->array

rapidsna commented on clang/test/CodeGen/builtin-counted-by-ref.c:
```diff
@@ -0,0 +1,182 @@
+  *_Generic(
+    __builtin_counted_by_ref(&p->array[0]),
+      void *: &__ignored,
+      default: __builtin_counted_by_ref(&p->array[idx++])) = size;
```
I see you have a test with a side effect. Though `idx++` is a parameter so the side effect would be optimized out in this O2 test anyway even if it was codegen'ed. Could you please use something that won't be optimized out to make sure the test is really checking it's not codegen'ed?

Also, is it intentionally to make `&p->array[non-zero]` work?

Sirraide replied:
Wait, why are we doing `-O2` in a test to begin with? Afaik codegen tests are usually supposed to run w/o optimisation; otherwise, you just end up testing the optimiser, and that’s not what Clang’s regression tests are for, after all.

rapidsna replied:
I agree.

bwendling replied:
I removed the `-O2` now. So the `idx` shouldn't be optimized away.

rapidsna replied:
Thanks! Do you have a test when the base has a side effect? The behavior seems to be different than `idx++` case, because when the base has a side effect the codegen will return `(void *)0` instead of returning the count reference. My inclination is to reject code or emitting a different warning message when the base has a side effect because the programmer might want to fix that instead of it being silently ignored or confused about the behavior.

bwendling replied:
If I do something like this:

```c
      *_Generic(
          __flex_counter((p++)->array),
              void *: &__ignored_assignment,
              default: __flex_counter(p->array)) = 42;
```

The code generation seems to take it in stride (with a warning of course):

```
define dso_local noalias noundef ptr @alloc_s(i32 noundef %size) local_unnamed_addr #0 {
entry:
  %0 = load ptr, ptr @p, align 8, !tbaa !5
  %..counted_by.gep = getelementptr inbounds i8, ptr %0, i64 4
  store i8 42, ptr %..counted_by.gep, align 1, !tbaa !9
  ret ptr null
}
```

I'm not sure if that's good or bad. In truth, the warning should probably be upgraded to an error, as it will lead to unexpected behavior if the warning is ignored. Thoughts?

rapidsna replied:
Yes, upgrading to an error sounds good to me! Also, the message might need to be updated (since the side effect is not actually ignored).

rapidsna commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -3691,6 +3691,38 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
+        const auto *FAMDecl = cast<FieldDecl>(ME->getMemberDecl());
+        if (const FieldDecl *CountFD = FAMDecl->findCountedByField()) {
+          Result = GetCountedByFieldExprGEP(Arg, FAMDecl, CountFD);
+          cast<llvm::GetElementPtrInst>(Result)->setNoWrapFlags(
```
Do you still need this?

bwendling committed:
Remote -O2 from codegen.

bwendling committed:
Remove setting the 'inBounds' flag.

bwendling commented:
Okay. I removed the `-O2` from the testcases. That should prevent optimizations from removing the `idx++`. I removed the frivolous setting if the InBounds. As for the `&ptr->array` question, if we're able to get the proper information from it I don't see why we can't have it :-). Anything else?

bwendling commented on clang/test/Sema/builtin-counted-by-ref.c:
```diff
@@ -16,6 +16,8 @@ void test1(struct fam_struct *ptr, int size, int idx) {
 
   *__builtin_counted_by_ref(ptr->array) = size;             // ok
   *__builtin_counted_by_ref(&ptr->array[idx]) = size;       // ok
+  *__builtin_counted_by_ref(&ptr->array) = size;            // ok
```
It looks like it generates correct code:

```c
  {
      unsigned long int __ignored_assignment;
      *_Generic(
          __flex_counter(&p->array),
              void *: &__ignored_assignment,
              default: __flex_counter(&p->array)) = 42;
  }

/*
define dso_local noalias noundef ptr @alloc_s(i32 noundef %size) local_unnamed_addr #0 {
entry:
  %0 = load ptr, ptr @p, align 8, !tbaa !5
  %..counted_by.gep = getelementptr inbounds i8, ptr %0, i64 4
  store i8 42, ptr %..counted_by.gep, align 1, !tbaa !9
  ret ptr null
}
 */
```

rapidsna replied:
I guess this will just work too? `*__builtin_counted_by_ref(ptr->array[idx]) = size;` 

Do you have a use case? I'd prefer these being ill-formed because the user might find it working unexpectedly. But I don't have a strong opinion on this. I'd leave it up to you/Linux users if you find this would be useful.

bwendling replied:
`*__builtin_counted_by_ref(ptr->array[idx]) = size;` doesn't work because the builtin requires a pointer to the FAM. (If you meant `&ptr->array[idx]`, that does work.)

I don't have any use case for `&ptr->array` or the like. I just don't want to be overly restrictive on what's considered a valid argument to the builtin. I think I'll do a comparison with gcc's version and see what they accept. We can match them.

bwendling replied:
GCC rejects everything that isn't an array. So not even `&ptr->array[0]`. I guess we can do that.

bwendling replied:
Note that, because GCC only checks if the argument is an array, it allows for things like:

```c
int global_array[];

void foo(int val) {
  *__builtin_counted_by_ref(global_array) = val;
}
```

which isn't great.

rapidsna replied:
I think we can still make it an error and have GCC to reflect our model on this one, unless Linux relies on this?

bwendling replied:
Yeah. I don't want to allow for global arrays either, just flexible array members.

bwendling committed:
Upgrade the side-effects warning into an error.

rapidsna commented on clang/test/CodeGen/attr-counted-by-pr88931.cpp:
```diff
@@ -13,7 +13,10 @@ void init(void * __attribute__((pass_dynamic_object_size(0))));
 // CHECK-LABEL: define dso_local void @_ZN3foo3barC1Ev(
 // CHECK-SAME: ptr noundef nonnull align 4 dereferenceable(1) [[THIS:%.*]]) unnamed_addr #[[ATTR0:[0-9]+]] align 2 {
 // CHECK-NEXT:  entry:
-// CHECK-NEXT:    tail call void @_Z4initPvU25pass_dynamic_object_size0(ptr noundef nonnull align 4 dereferenceable(1) [[THIS]], i64 noundef -1) #[[ATTR2:[0-9]+]]
```
`_Z4initPvU25pass_dynamic_object_size0 ` is not shown here anymore. Does it still check what it needs to check?

bwendling replied:
The -O2 flag was removed. That means the `__builtin_object_size` stuff won't be evaluated.

rapidsna replied:
Yes, that's what I figured. I guess we still need `-O2` for tests that involve `__builtin_object_size`? @Sirraide WDYT?

Sirraide replied:
> Yes, that's what I figured. I guess we still need `-O2` for tests that involve `__builtin_object_size`? @Sirraide WDYT?

No idea candidly. I’m not familiar w/ how that builtin works. I’d recommend moving the tests for this to a separate file in that case.

bwendling replied:
What do you mean put in a separate file? This *is* the separate file. :-)

Sirraide replied:
Wait, I didn’t notice that this file has nothing to do w/ `__builtin_counted_by_ref` (or am I missing something here?). Because if not, why did you remove the `-O2` from this file?

Sirraide replied:
Essentially, what I meant was that there is no reason to ever write a codegen test w/ optimisation enabled—unless the code that you’re introducing has different behaviour depending on optimisation levels (which iirc isn’t the case for `__builtin_counted_by_ref`, right?).

*Other* tests might still require the `-O2`; and if that is the case, then we generally shouldn’t add tests for something that is invariant under optimisation (wrt how the frontend behaves; the IR would be different of course, but then has nothing to do w/ what Clang emits) to a file that contains tests that *do* care about optimisation levels. I thought at first that this is what you did here, but looking at it again, that doesn’t seem to be the case because there *is* in fact a separate file w/ tests for this pr.

So yeah, my bad if I wasn’t too clear about this ealier.

Sirraide replied:
> different behaviour depending on optimisation levels

As in, Clang emits different IR before it even gets to the optimiser.

rapidsna replied:
I also don't see why the other tests need to be updated here, except if you have to remove test cases that no longer works with new restrictions. Still, the rest of the tests and file checks should stay the same.

bwendling replied:
I removed those tests from this PR. PTAL.

Sirraide replied:
That’s better yeah; they seem completely unrelated, so we shouldn’t really be messing w/ them here.

rapidsna commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -4914,3 +4914,9 @@ def ArithmeticFence : LangBuiltin<"ALL_LANGUAGES"> {
+def CountedByRef : Builtin {
+  let Spellings = ["__builtin_counted_by_ref"];
+  let Attributes = [NoThrow, CustomTypeChecking];
+  let Prototype = "int(...)";
```
Just to make sure. I guess the prototype doesn't match (`void *` vs `T *`). Is it still fine?

bwendling replied:
For a custom type checking, the prototype is basically meaningless. I'd rather not mess with this as it's not very important.

bwendling committed:
Restrict __builtin_counted_by_ref to 'ptr->array' to match GCC's implementation.

bwendling committed:
Regenerate.

bwendling committed:
Revert the modifications as they're not affected by this feature.

rapidsna commented on clang/test/Sema/builtin-counted-by-ref.c:
```diff
@@ -0,0 +1,122 @@
+void test3(struct fam_struct *ptr, int idx) {
+  __builtin_counted_by_ref(&ptr->array[idx]);               // expected-error {{'__builtin_counted_by_ref' argument must reference a flexible array member}}
+  __builtin_counted_by_ref(&ptr->array);                    // expected-error {{'__builtin_counted_by_ref' argument must reference a flexible array member}}
+  __builtin_counted_by_ref(ptr->x);                         // expected-error {{'__builtin_counted_by_ref' argument must reference a flexible array member}}
```
Nit: we can also add `__builtin_counted_by_ref(ptr->array[0]);`

rapidsna commented:
I still see `CodeGeb/attr-counted-by*.c` being unnecessarily changed. Should we just revert it back to `llvm:main`. Other than a couple of minor comments on the tests, LGTM! Thanks for addressing feedback. You might still wait on @AaronBallman though.

Sirraide commented:
> I still see `CodeGen/attr-counted-by*.c` being unnecessarily changed.

Yeah, the `-O2` definitely needs to be added back to those tests.

bwendling committed:
Add testcase.

bwendling committed:
Revert.

bwendling commented:
> > I still see `CodeGen/attr-counted-by*.c` being unnecessarily changed.
> 
> Yeah, the `-O2` definitely needs to be added back to those tests.

Fine, but I'm not touching anymore testcases again. I'm getting quite sick of you two ping-ponging back and forth on already good testcases.

bwendling committed:
Revert unaffected tests.

bwendling committed:
Merge branch 'llvm:main' into builtin_get_counted_by

bwendling committed:
Fix Sphinx formatting.

nikic commented:
This change causes a large compile-time regression: https://llvm-compile-time-tracker.com/compare.php?from=ae9d0623ad65d84022bb4ed8446b6491451ae575&to=7475156d49406785a974b1205d11fe3de9c1553e&stat=instructions:u

From a quick look, I'd guess it's due the extra AST traversals you have added in SemaExpr.

bwendling commented:
I'll check it out. Sorry about the regression.
